crates/isograph_compiler/src/source_files.rs

```diff
@@ -0,0 +1,314 @@
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    ops::{Deref, DerefMut},
+    path::{Path, PathBuf},
+};
+
+use common_lang_types::{SourceFileName, TextSource};
+use graphql_lang_types::{GraphQLTypeSystemDocument, GraphQLTypeSystemExtensionDocument};
+use graphql_schema_parser::{parse_schema, parse_schema_extensions};
+use intern::string_key::Intern;
+use isograph_config::CompilerConfig;
+use isograph_lang_parser::IsoLiteralExtractionResult;
+use isograph_schema::UnvalidatedSchema;
+
+use crate::{
+    batch_compile::BatchCompileError,
+    isograph_literals::{read_and_parse_iso_literals, read_file, read_files_in_folder},
+    schema::read_schema_file,
+    watch::{ChangedFileKind, SourceEventKind, SourceFileEvent},
+};
+
+#[derive(Clone, Debug, Eq, PartialEq, Default)]
+pub struct SourceFiles {
+    pub schema: GraphQLTypeSystemDocument,
+    pub schema_extensions: HashMap<SourceFileName, GraphQLTypeSystemExtensionDocument>,
+    pub contains_iso: ContainsIso,
+}
+
+impl SourceFiles {
+    pub fn is_empty(&self) -> bool {
+        self.schema.0.is_empty()
+            && self.schema_extensions.is_empty()
+            && self.contains_iso.is_empty()
+    }
+
+    pub fn read_and_parse_all(&mut self, config: &CompilerConfig) -> Result<(), BatchCompileError> {
+        // Schema
+        self.schema = read_and_parse_graphql_schema(&config.schema)?;
```

Comment:
we should probably be creating an entirely new `SourceFiles` here and calling `*self = new_source_files`.

For example, right now, we don't clear out the `schema_extensions` and `contains_iso`.

Here, I am thinking of someone else making a change in the future (e.g. adding a field to `SourceFiles`) and this code note breaking.